Add initial troubleshooting guide (#1430)
kubernetes-sigs/gateway-api-inference-extension: site-src/guides/troubleshooting.md

> # Toubleshooting Guide

Reviewer: ```suggestion
# Troubleshooting Guide
```

This guide provides troubleshooting steps and solutions for common issues encountered with the Gateway API inference extension.

## 400 Bad Request

### `model not found in request body` or `prompt not found in request`

> If the OpenAI API endpoint you're using isn't working as expected, the issue might be related to the request body format. The endpoint picker (EPP) assumes that if a request is a POST, its body must contain the `model` and `prompt` fields. This is because the gateway currently assumes the requests are for Large Language Models (LLMs).

Reviewer: ```suggestion
If the OpenAI API endpoint you're using isn't working as expected, the issue might be related to the request body format. The endpoint picker (EPP) assumes that if a request is a POST, its body must contain the `model` field and the `prompt` (or `messages`) field. This is because the gateway currently assumes the requests are for Large Language Models (LLMs).
```

Reviewer: This is on me, I forgot this functionality was added. Source:https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/pkg/epp/util/request/body.go#L25

**Solution**: Make sure your request body contains the missing field.

## 404 Not Found
This is a default gateway error, meaning the request never reached a backend service. This usually means that there is no HTTPRoute configured to match the request path (e.g. /v1/completions). The gateway doesn't know where to send the traffic.

**Solution**: Ensure you have an HTTPRoute resource deployed that specifies the correct host, path, and backendRef to your InferencePool.

## 429 Too Many Requests
### `system saturated, sheddable request dropped`

> This error indicates that the entire request pool has exceeded its saturation thresholds. This means the system is under heavy load and is shedding non-critical requests. To address this, check the following:

Reviewer 2: ```suggestion
This error indicates that the entire inference pool has exceeded its saturation thresholds. This means the system is under heavy load and is shedding non-critical requests. To address this, check the following:
```

    * `DefaultKVCacheUtilThreshold`: This is the maximum utilization of the Key-Value (KV) cache on the model server, expressed as a decimal from 0.0 to 1.0. The default is 0.8, or 80%. The KV cache stores attention keys and values to speed up inference for subsequent tokens. When its utilization exceeds this threshold, it's an indication that the model server is nearing its memory capacity and may be becoming saturated. To override this, set the `SD_KV_CACHE_UTIL_THRESHOLD` environment variable.
    * `DefaultMetricsStalenessThreshold`: This defines the maximum age of metrics data before it's considered outdated. The default is 200 milliseconds. The saturation detector needs up-to-date metrics to make accurate decisions about system load. If the metrics are older than this threshold, the detector won't use them. This value is tied to how often metrics are refreshed, and setting it slightly higher ensures that there's always fresh data available. To override this, set the `SD_METRICS_STALENESS_THRESHOLD` environment variable.

## 500 Internal Server Error

> ### `fault filter abort`

Reviewer 2: Are these errors specific to GKE gateway or are they common Gateway API error codes?

Author: That's a good question. The 500 fault filter abort error is common when the backend does not exist. The 503 is also common when there is no healthy endpoint to route to. The only thing that can diverge is when port is misconfigured - I believe GKE and Istio gives 503 Service Available but KGateway gives 502 Bad Gateway. I have updated the doc to add 502 error code.

**Solution**: Create or update the RBAC configuration to grant the [required permissions](https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/137a0b4660b96487caac626ed135b3600be876ed/config/manifests/inferencepool-resources.yaml#L129) to the EPP service account.

### `Pool is not initialized, skipping refreshing metrics`
This error indicates that the Inference Pool pods are not initialized. 

**Solution**: Check the EPP start up argument `--pool-name` has the correct InferencePool name specified and the InferencePool exists.

## Unexpected Routing Behaviors
The EPP's core function is to intelligently route requests to the most optimal model server pod in a pool. It uses a score-based algorithm that considers several metrics (such as queue depth, KV cache utilization, etc.) to choose the best pod for each request. 

For unexpected routing behaviors: 

> * Verify the expected metrics are being emitted from the model server. Some model servers aren't fully compatible with the default expected metrics, vLLM is generally the most up-to-date in this regard.

Reviewer 2: Pls add the supported model server link https://gateway-api-inference-extension.sigs.k8s.io/implementations/model-servers/

## 500 Internal Server Error
### `fault filter abort`
This internal error suggests a misconfiguration in the gateway's backend routing. Your HTTPRoute is configured to point to an InferencePool that does not exist or cannot be found by the gateway. The gateway recognizes the route but fails when trying to send traffic to the non-existent backend.

**Solution**: Verify that the backendRef in your HTTPRoute correctly names an InferencePool resource that is deployed and accessible in the same namespace. If you wish to route to an InferencePool in a different namespace, you can create a `ReferenceGrant` like below:

```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: ReferenceGrant
metadata:
  name: ref-grant
  namespace: ns2
spec:
  from:
    - group: gateway.networking.k8s.io
      kind: HTTPRoute
      namespace: ns1
  to:
    - group: inference.networking.k8s.io
      kind: InferencePool
      name: my-inference-pool

> ```

Reviewer 3: out of curiosity, was this ever tested?

Author: Not sure about other implementations but recently we found a bug where ReferenceGrant does not work properly with InferencePool, that was fixed in GKE.

## 503 Service Unavailable
### `no healthy upstream`
This error indicates that the HTTPRoute and InferencePool are correctly configured, but there are no healthy pods in the pool to route traffic to. This can happen if the pods are crashing, still starting up, or failing their health checks.

**Solution**: Check the status of your model server pods. Investigate the pod logs for any startup errors or health check failures. Ensure your model server is running and listening on the correct port and that any configured healthchecks / readiness probes are succeeding.

## The endpoint picker (EPP) Crashlooping
When EPP is crashlooping, check the logs of your EPP pod. Some common errors include:

### `failed to list <InferencePool or InferenceObjective or Pod>: … is forbidden`
The EPP needs to watch the InferencePool, InferenceObjectives and Pods that belong to them. This constant watching and reconciliation allows the EPP to maintain an up-to-date view of the environment, enabling it to make dynamic decisions. This particular error indicates that the service account used by the EPP doesn't have the necessary permissions to list the resources it’s watching.

> **Solution**: Create or update the RBAC configuration to grant the [required permissions](https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/137a0b4660b96487caac626ed135b3600be876ed/config/manifests/inferencepool-resources.yaml#L129) to the EPP service account.

Author: Thanks for pointing this out! I will create a follow up CL to fix this.
